### PR TITLE
Added a search method for the author and title parameters

### PR DESCRIPTION
**Pull request by Hryniv**

### CODE REVIEW

File: src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import com.bookstore.model.Book;
+import com.bookstore.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java]: create constant

File: src/main/java/com/bookstore/repository/book/BookSpecificationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private final SpecificationProviderManager<Book> specificationProviderManager;
+
+    @Override
+    public Specification<Book> build(BookSearchParametersDto searchParameters) {
+        Specification<Book> sp = Specification.where(null);
+        if (searchParameters.author() != null) {
+            sp = sp.and(specificationProviderManager.getSpecificationProvider("author")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/repository/book/BookSpecificationBuilder.java]: create constants

File: src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import com.bookstore.model.Book;
+import com.bookstore.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java]: create constant

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    private static final String AUTHOR = "author";
+
+    @Override
+    public String getKey() {
+        return AUTHOR;
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String param) {
+        return (root, query, criteriaBuilder) -> root.get("author").in(param);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/repository/book/AuthorSpecificationProvider.java]: use your constant

File: src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    private static final String TITLE = "title";
+
+    @Override
+    public String getKey() {
+        return TITLE;
+    }
+
+    @Override
+    public Specification<Book> getSpecification(String param) {
+        return (root, query, criteriaBuilder) -> (root.get("title").in(param));
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/repository/book/TitleSpecificationProvider.java]: same

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -46,6 +50,15 @@ public BookDto updateBook(Long id, CreateBookRequestDto requestDto) {
         return bookMapper.toDto(book);
     }
 
+    @Override
+    public List<BookDto> search(BookSearchParametersDto searchParameters) {
+        Specification<Book> spec = bookSpecificationBuilder.build(searchParameters);
+        return bookRepository.findAll(spec)
+                .stream()
+                .map(bookMapper::toDto)
+                .toList();
```
Comment (okuzan) [inline src/main/java/com/bookstore/service/impl/BookServiceImpl.java]: FYI you can declare mapping list to list in mapstruct too with one line

File: src/main/java/com/bookstore/repository/book/BookSpecificationProviderManager.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> specificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return specificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
Comment (okuzan) [inline src/main/java/com/bookstore/repository/book/BookSpecificationProviderManager.java]: change p to provider

**Review by okuzan — APPROVED** [review]

**Pull request merged**